artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java

```diff
@@ -517,7 +517,6 @@ public void removeRolledback(MessageReference messageReference) {
    }
 
    public void addRolledback(MessageReference messageReference) {
-      currentWindow.decrementAndGet();
```

Comment:
why? were you having issues here?

Reply:
It was a while since I looked at that, but I think this change will enable the consumer to fetch messages past the prefetchLimit. It happens when doing redeliveries, such that it can theoretically prefetch twice as many messages as it's supposed to.

Again, saw this quite some time back so I could be mistaken...

Reply:
...yes, that is indeed the case. Verified by adding messages to rollback in "testRedeliveredMessageNotOverflowingPrefetch" and printing deliveryCount instead of asserting.

Reply:
@gtully unless adressed elsewhere, this will still cause a regression on the openwire consumers prefetch-limit...

Reply:
will investigate a bit more here, thanks

Reply:
in testRedeliveredMessageNotOverflowingPrefetch: currentWindow goes negative and message 14 does not arrive! the logic that puts messaged back on the delivered list and pulls from the transaction is in the mix here. That list has 11 messages, prefetch + the one rolled back message. I think currentWindow should only be decremented by the 1 rolledback message. 
This is very tricky, but it the root problem may be that there is a single transaction now, doing the rollback of the client message and the prefetched messages.

Reply:
@AntonRoskvist I pushed a new test that fails with a revert of line 520, the decrement. I don't yet know what is at fault, but the root cause is the combination of rolledback and delivered messages in a single transaction. 
we decrement 11 times in stead of 1 if the session transaction is in play. 
it may be that cancelRefs needs its own transaction all the time, the comments at: https://github.com/apache/activemq-artemis/blob/6e28e4683458490ab0e3990ba8f35e9c6531c461/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerSessionImpl.java#L2234
may have a hint - it speaks to the same problem.
it may be that cancleRefs should not have a tx at all, or a new one all the time that has no IO may work ok.

Reply:
Quite interesting... I have actually noticed queue counters (metrics) being incorrect from time to time on queues using OW consumers... in particular ones using XA-transactions. This type of application is getting exceedingly rare in the envs I manage, and it has not happened often enough to warrant further investigation, but it is something I have seen multiple times. Guessing it could have a common source to this as well.

For the added test you mentioned though, I found it to behave pretty much as I would expect it to, besides prefetch being of-by-one (which I guess might be expected given the rollback, not sure). 

The error I am getting is caused by a null value returned from the receive-call, which happens because some message gets moved over to DLQ after exceeding "maximumRedeliveries"... so this is the expected result given the queue is empty.

Setting `connection.getRedeliveryPolicy().setMaximumRedeliveries(-1);` makes the test pass.

Additionally I ran with `queueControl.getDeliveringCount()` printed for each loop iteration and prefetch never exceeds prefetchPolicy + 1.


I'm not sure about this next part at all: 
From what I recall from last time I looked at this, I think I interpreted the incorrect `currentWindow` and `prefetch` values to be caused by from some difference in how OW handled redeliveries compared to Core... something about the OW client holding on to the message for each redelivery but also notifying the broker each time about the redelivery, whereas for Core the message actually "bounces" back and forth for each delivery.

Reply:
Thanks. There was something weird, but messages going to the DLQ was part of it.
The other part is the returnReferences that puts messages back in delivering rather than back on the queue on a tx rollback. With the core session in play, that also had prefetched messages in the tx, and that was breaking the returnReferenced logic and causing to many decrements on the currentWindow. 
Thanks for picking up on that change :-)

Reply:
I played with doing the session cancelRef in a new transaction, but multiple transactions can interact with IO and become out of sync due to any operation context task clearing the current context. When the delivery count is persisted, there is IO. That was to root cause of the original investigation, so I don't want to introduce that possibility again. I think what i have now is good.